OpenHMD for Godot 3.0

BastiaanOlij opened the pull request:
This is a first pass at implementing the open source OpenHMD project into Godot based ontop of my ar_vr_server architecture.

I've taken a few shortcuts so there are a few improvements needed but for the most part this seems to work.

OpenHMD is an open source project that aims to bring cross platform VR headset support without the need for external drivers.
At this point in time OpenHMD only supports basic orientation tracking and no positional tracking but it does work with Rift, Vive and PSVR amongst others.

Output is to the main Godot window so it assumes you're mirroring your display and are running your application full screen.

BastiaanOlij committed:
OpenHMD VR interface

BastiaanOlij commented:
@reduz it would be good if you could have a look at my serious hacking attempts to directly use the lens distortion shader for output. I want to add something similar to my mobile VR implementation in the base ar/vr implementation. Pretty sure this is not the right way to implement it but need to pick your brain on how this should be done to make use of the existing internals of Godot.

BastiaanOlij commented:
Anyone with more Linux experience then me, please have a look at running this if you have a Vive, Rift or PSVR. 

As I understand it there are two versions of hidapi you can use, one based on hidraw which requires udev, and one based on libusb which is not part of the build process yet.

The main issue that we seem to be running into is a permissions issue which may, or may not be solved using the libusb version. 

There is more info on the issue in these two files:
https://github.com/signal11/hidapi/blob/master/linux/README.txt
https://github.com/signal11/hidapi/blob/master/udev/99-hid.rules

BastiaanOlij commented:
OK this last check in is very experimental. In attempting to resolve the issues on linux I've added a copy of libusb but lacking a linux environment to test on, this may not work.

The windows version seems to be working well, waiting on some fixes on the PSVR side of life. 
On mac it should work for the most part too.

BastiaanOlij commented:
@akien-mga Travis falls over on the usage of some c99 stuff in the OpenHMD drivers. What is our current stance on enabling c99? I don't know how much is in any of the other files but so far seems to be inline declaration of variables in for loops that it falls over, it would be easy enough to submit a PR to work around that if the OpenHMD guys don't mind, but seeing this won't be used outside of the Mac/Windows/Linux sphere may be easier to just enable it.

I'm assuming it builds on Linux because of the udev exclusion (which I'm about to remove as I'm trying to switch to using libusb) and on Mac because xcode probably allows c99 by default.

Btw, are you able to test building this for me on Linux, I'm going completely blind here.

BastiaanOlij commented on servers/visual/visual_server_viewport.cpp:
```diff
@@ -260,15 +265,25 @@ void VisualServerViewport::_draw_viewport(Viewport *p_viewport) {
+	Ref<ArVrInterface> arvr_interface = ArVrServer::get_singleton()->get_primary_interface();
+	if (arvr_interface.is_valid()) {
+		// update our positioning information as late as possible...
+		float delta = 0; ///@TODO need to set this, time accurately since last time we called this!
```
@reduz is there anything I can call here to determine how much time has passed since the last time this was called? I noticed there where some timing functions in OS but I couldn't tell if they were implemented for all platforms.

TheOnlyJoey commented:
This patch will fix the libusb compile, you missed setting the OS defines.
Fixes the Linux permission issues.

```
diff --git a/modules/openhmd/SCsub b/modules/openhmd/SCsub
index 6a6bdf17b..e71a18f1a 100644
--- a/modules/openhmd/SCsub
+++ b/modules/openhmd/SCsub
@@ -35,16 +35,19 @@ if p == 'x11':
         env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/linux_usbfs.c" ])
         env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/poll_posix.c" ])
         env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/threads_posix.c" ])
+        env_openhmd.Append(CPPDEFINES=["OS_LINUX"])
 #    elif p == 'windows':
 #        env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/windows_nt_common.c" ])
 #        env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/windows_usbdk.c" ])
 #        env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/windows_winusb.c" ])
 #        env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/poll_windows.c" ])
 #        env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/threads_windows.c" ])
+#        env_openhmd.Append(CPPDEFINES=["OS_WINDOWS"])
 #    elif p == 'osx':
 #        env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/darwin_usb.c" ])
 #        env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/poll_posix.c" ])
 #        env_openhmd.add_source_files(env.modules_sources, [ "#thirdparty/libusb/libusb/os/threads_posix.c" ])
+#        env_openhmd.Append(CPPDEFINES=["OS_DARWIN"])
 
 # Link in hidapi
 hidapi_headers = "#thirdparty/hidapi/hidapi/"
```

BastiaanOlij commented:
@TheOnlyJoey I am not surprised I missed those, I shall add them tonight! Thanks man!

I'll start experimenting with using libusb on Mac as well. I'm not sure if it will work on Windows as I had issues with hidapi including POSIX based things that it didn't seem to like very much (libusb itself seemed to compile) but maybe on Mac it will give us device names that are more in line with what we're seeing on Linux?

I've also asked a mate if I can borrow his Vive to see how well that works. I've not had the Vive able to run on either of my laptops because OpenVR refuses to work with the GPU I have (I think the GPU is actually find but the HDMI output is limited to 60Hz) and it seems an artificial limitation just to ensure you're not running on hardware that might induce more motion sickness due to the inherent frame rate limit. I'd be keen to see if this all runs without that restriction even if it is only good for prototyping.

Did it all run fine now on Linux?

BastiaanOlij commented:
Ok, this latest update is mostly just refactoring code. I've moved most of the code related to the lens distortion shader into it's own files encapsulated in a new class called OpenHMDShader. 

@reduz I'd like you're opinion on that class, the shader code itself comes from OpenHMD but the class contains all the logic to compile and use it. There is an argument that more belongs on the OpenHMD side, there is also an argument that I'm duplicating logic that already exists in Godot and really belongs in the GLES3 drivers.

All in all it means the OpenHMD interface itself has become pretty clean and really only contains the "glue" between Godot and OpenHMD.

Time will tell how much still needs to change once controller support and positional tracking are added to OpenHMD, one idea I've been toying with is to expose each device found by OpenHMD as a separate instance to make the UI more natural, but equally I'm thinking that if you build a VR game you'd build it against the interface of your choice.

Anyway, it's late enough as it is. @akien-mga I still need your input on Travis complaining about c99.

akien-mga commented:
@BastiaanOlij Add this to `modules/openhmd/SCsub`:
```
# upstream uses c99
env_openhmd.Append(CFLAGS="-std=c99")
```
```

beniwtv commented:
Two questions here @BastiaanOlij (if you don't mind), as I would be interested to try this on Linux as well (I have a Vive):

1. I saw mention in this PR about Linux - Is this related to SteamVR at all? Or can it run without it? I ask because according to Valve, only Vulkan is supported for VR on Linux, OpenGL is too slow. If it runs without - anything to set-up manually, or will OpenHMD take care of, say, enabling the headset (or waking it up from sleep)?

2. You mention it assumes displays are mirrored - by default though the Vive normally runs as extended (or direct mode, but that may or may not work currently). Games will normally make a normal window (without distortion) + another one on the extended for the HMD. Is that something planned for later?

Many thanks and awesome work!

BastiaanOlij commented:
Hey @beniwtv , I don't have a linux machine myself so I'm a bit handycapped at commenting however note I have two separate PRs at this point in time.

This PR which is based on the OpenHMD open source project which implements the Oculus Rift, HTC Vive, PSVR and a few other HMDs. They are able to directly interact with the hardware for tracking but they leave the visual side of things up to our side of the fence. At this point display mirroring is the only option we have available in this PR and this is a part where much more work is required.

Now I have to admit I do not know enough about what the smart people at Valve have seen as roadblocks, they do this 8+ hours a day 5 days a week, I'm trying my best to keep up in my spare time though the guys working on the OpenHMD framework are putting far more time into this and may be better capable of answer those questions. Note that the OpenHMD guys do most of their development on Linux so obviously it all works.

One limit I have run into personally is that the display mirroring can limit the framerate to the lowest capable device which would be the normal monitor. The architecture in the ARVR server supports what mention in your last point but it's not implemented in this PR. I'll come back to that in a minute.

How much Vulkan has a play in this I do not know at this point in time. I know they had to do some magic on the Mac side of life with additions to Apples Metal as well but I think a lot has to do with actually accessing the display, that in order to actually output the frames generated by the rendering engine at a high enough throughput, they have to go beyond what OpenGL delivers in its standard. 

I'm totally guessing here though but very simply put, Sony is able to crank out 90fps+ from hardware far less beefy then what you find on most PCs and most Macs regardless of what OS you run. The same goes for the new HMDs Microsoft is pushing, they are designed to work with relatively low spec hardware (at the expense of quality, MS doesn't care for next gen looking graphics they care for a more functional approach), the roadblock I hit on my MacBook Pro and HP Probook is that they can't output an HMDI signal at anything over 60hz, so the Vive and Rift refuse to work on their native SDKs. I think this is a common issue people will run into who are using laptops unless you have a gaming laptop that supports newer HDMI standards. 

I've seen plenty of games developed for the Vive that do all their rendering in OpenGL with stock standard GLSL shaders, some even using the old OpenGL 1/2 rendering pipeline. But they do use the OpenVR framework (and thus go over SteamVR) and I think the only 'magic' here is that its internals use some of the newer graphics APIs to access hardware more directly that allows them to run things at a higher throughput (and do some funky things like timewarp and stuff to handle any framerate drops in your game and insert constructed frames).

Which after a long rant brings me to my other PR which implement the OpenVR SDK as is and while there are still a few footnotes to work out it pretty much fully functional.

Regardless of whether you use OpenHMD or OpenVR or anything else, the ARVR server renders each frame once for each eye as a normal flat image. It is the VR implementation that applies the lens distortion and outputs to the HMD + screen. in OpenHMD its currently just applies and output to screen (hence the need for mirroring at the moment), for OpenVR it will sent each image to the OpenVR SDK for further processing, OpenVR applies the lens distortion and outputs to the HMD and applies any other post processing needed such as chaparone and such. 

If you turn the AR/VR toggle on for the main viewport (as the GodotTestARVR project does) the left eye image is also blitted out to screen as it is. 

Alternatively you can add a viewport to your project and turn ar/vr mode on on that one. This will cause that viewport to render to the HMD but nothing will be rendered to screen. You can thus use a camera for the HMD independent to the camera your render to screen and create a co-op game :)

One problem with this that I haven't dealt with is that I want to limit the FPS for this blit so the HMD renders at full framerate but the screen outputs at a lower rate because its just wasted processing.

How this all works for OpenVR on MacOS or Linux? I'm not sure. I know @Hinsbart had the linux version of OpenVR working but he did mention it very much depended on whether the correct GFX drivers were available. For MacOS it obviously requires High Sierra so you have access to metal2, one of the new iMacs or an external GPU to run.

Anyway, to make a long story short, you should be able to do what you want to do today with the OpenVR PR. OpenHMD should run as well but I do believe there is far more work to do to handle the output better.

beniwtv commented:
Oh, I see, thanks for the clarifications, will play around a bit with this and the other PR ;) @BastiaanOlij

akien-mga commented on platform/windows/detect.py:
```diff
@@ -190,7 +190,7 @@ def configure(env):
             env.Append(CCFLAGS=['/D_WIN64'])
 
-        LIBS = ['winmm', 'opengl32', 'dsound', 'kernel32', 'ole32', 'oleaut32', 'user32', 'gdi32', 'IPHLPAPI', 'Shlwapi', 'wsock32', 'Ws2_32', 'shell32', 'advapi32', 'dinput8', 'dxguid']
+        LIBS = ['winmm', 'opengl32', 'dsound', 'kernel32', 'ole32', 'oleaut32', 'user32', 'gdi32', 'IPHLPAPI', 'Shlwapi', 'wsock32', 'Ws2_32', 'shell32', 'advapi32', 'dinput8', 'dxguid', 'Setupapi']
```
Linking fails while cross-compiling from Unix:
```
x86_64-w64-mingw32-g++ -o bin/godot.windows.tools.64.exe -static platform/windows/godot_win.windows.tools.64.o platform/windows/context_gl_win.windows.tools.64.o platform/windows/os_windows.windows.tools.64.o platform/windows/ctxgl_procaddr.windows.tools.64.o platform/windows/key_mapping_win.windows.tools.64.o platform/windows/tcp_server_winsock.windows.tools.64.o platform/windows/packet_peer_udp_winsock.windows.tools.64.o platform/windows/stream_peer_winsock.windows.tools.64.o platform/windows/joypad.windows.tools.64.o platform/windows/power_windows.windows.tools.64.o platform/windows/godot_res.windows.tools.64.o main/libmain.windows.tools.64.a main/tests/libtests.windows.tools.64.a modules/libmodules.windows.tools.64.a drivers/libdrivers.windows.tools.64.a editor/libeditor.windows.tools.64.a scene/libscene.windows.tools.64.a servers/libservers.windows.tools.64.a core/libcore.windows.tools.64.a modules/freetype/libfreetype_builtin.windows.tools.64.a -lmingw32 -lopengl32 -ldsound -lole32 -ld3d9 -lwinmm -lgdi32 -liphlpapi -lshlwapi -lwsock32 -lws2_32 -lkernel32 -loleaut32 -ldinput8 -ldxguid -lSetupapi
/usr/bin/x86_64-w64-mingw32-ld: cannot find -lSetupapi
collect2: error: ld returned 1 exit status
```

I guess it should be `setupapi`:
```
# find /usr/i686-w64-mingw32/ | grep -i setupapi
/usr/i686-w64-mingw32/sys-root/mingw/lib/libsetupapi.a
/usr/i686-w64-mingw32/sys-root/mingw/include/setupapi.h
```

BastiaanOlij commented:
rebased.

neikeq commented on servers/arvr_server.cpp:
```diff
@@ -0,0 +1,313 @@
+	ClassDB::bind_method(D_METHOD("get_interface:ARVRInterface", "idx"), &ARVRServer::get_interface);
+	ClassDB::bind_method(D_METHOD("find_interface:ARVRInterface", "name"), &ARVRServer::find_interface);
+	ClassDB::bind_method(D_METHOD("get_tracker_count"), &ARVRServer::get_tracker_count);
+	ClassDB::bind_method(D_METHOD("get_tracker:ARVRPositionalTracker", "idx"), &ARVRServer::get_tracker);
```
Starting today, the `:type` suffix must not be used anymore, except for virtual method binds.

BastiaanOlij commented:
Again, just rebased :)

beniwtv commented:
So got some time to test this now, tracking seems to be working fine (except "positional tracking", but that isn't supported by OpenHMD for the Vive right now). 

Once I moved the window to the right screen and maximized it, it showed up on the HMD fine, but it seems that lens distortion isn't working. Not sure if that is something OpenHMD should handle, or if need to do this manually in the app.

(Side note here about the "framerate while mirroring"-problem: Maybe the solution here is to assume extended mode instead, and have Godot figure out which "screen" the HMD display is and move the window there automatically.)

I also couldn't get controller tracking working, which probably depends on getting the tracking working for the Vive in the first place in OpenHMD.

BastiaanOlij commented:
Hey @beniwtv , 

OpenHMD currently has no tracking for controllers or positional tracking for the headset. That is coming for the vive, I think Joey said they are planning to merge master in a few weeks time.

It should be applying lens distortion but again it may be something where the settings haven't ended up in the OpenHMD master yet.

The whole windowing thing is my biggest beef at the moment with the OpenHMD implementation. It is purposely not part of the OpenHMD implementation so we're taking care of that ourselves (or better said, not taking care of it at all :)). Joey mentioned that they are working on a couple of support projects that should make it easier to add the required functionality, I really don't want to spend the time reinventing the wheel here especially lacking access to hardware to get it to work properly :)

BastiaanOlij commented:
@TheOnlyJoey I've just rebased this to todays master, I've also undone my changes to OpenHMD and instead moved the shaders back into openhmd_shader.cpp so that I can use OpenHMD as it is.
I've also added in the NOLO stuff with that small change to make it compile without relying on a c99 capable compiler (I'm on windows bah).. It is however the source as at the start of this week and I saw that  you checked in a few changes into the NOLO branch.

I have not applied any of your changes regarding NOLO tracking or controllers. I'm going to start looking at adding them in a bit nicer on Saturday. I've also asked Peter if I can borrow his NOLO and he's not using it atm so with luck I can actually do some proper testing.

DO copy this out into a new folder to test it before overwriting what you're working on. I'm getting one of the background threads crashing when the project selector comes up. It doesn't seem to stop anything from running and might just be a fluke related to my computer or that I'm compiling everything with a skeleton install of VC++ 2017, dunne know...

BastiaanOlij commented:
I've added a whole bunch of untested changes so don't overwrite your current working version if you're playing with this until I hear back from some people testing this stuff. Hopefully I'll have some devices next week to do some testing myself.

There are a few things that have been renamed and a few things that have been fixed.

There is now a new function you can call on the interface called init_tracking_device. By default the positional and rotational tracking of the initialized HMD device is used but this allows you to use a different device. Right now this is used to use the NOLO HMD puck to track the HMD when you are using the NOLO system but there are other use cases as more and more tracking solutions hit the market.

I have also added a function called init_controller_device. This is mostly a temporary solution so you can initialize the new NOLO controller devices. This is far from finished logic, just a first go at it really. Ideally I would like this to auto detect controllers that have been turned on so the tracking is automatically enabled and you just get a signal just like we do in the OpenVR branch. But for now its a manual add.

BastiaanOlij commented:
Rebased and applied clang-format

BastiaanOlij commented:
Rebased, added latest openhmd master with nolo support (and a small change to work around unsupported c99 compile issue on windows), and fixed issue with program not being restored.

BastiaanOlij commented:
Rebased and a few small tweaks. I'll be starting on porting this to our GDNative solution now so this is possibly going to be the last rebase on this PR.

@TheOnlyJoey , did you guys ever did something about the usage of c99?

BastiaanOlij commented:
Rebased again but didn't do a whole lot of testing. This is now both in sync with todays master of Godot 3.0 and todays master of OpenHMD

BastiaanOlij commented:
This has been superseeded by the GDNative OpenHMD driver: https://github.com/GodotVR/godot_openhmd
